Added validation flow with context

kobmic opened the pull request:
Added possibility to add a sequence of contexts containing supported TransferSyntaxUID, MediaSOPClassUID/SOPClassUID combinations.
- if context is given validation flow will try to buffer 512 bytes for validation, otherwise 140.
- if byte stream contains DICOM file meta information, TransferSyntaxUID and MediaSOPClassUID are required and compared to supported contexts
- if no file meta information is present, TransferSyntax will be "guessed" and SOPClassUID is required and compared to supported contexts

KarlSjostrand commented on src/main/scala/se/nimsa/dcm4che/streams/DicomFlows.scala:
```diff
@@ -41,14 +43,20 @@ object DicomFlows {
+  case class Context(sopClassUID: String, transferSyntax: String)
+
+
+  private class DicomValidateFlow(contexts: Option[Seq[Context]]) extends GraphStage[FlowShape[ByteString, ByteString]] {
```
This class is growing. Consider moving it to its own file.

KarlSjostrand commented on src/main/scala/se/nimsa/dcm4che/streams/DicomFlows.scala:
```diff
@@ -62,34 +70,167 @@ object DicomFlows {
 
         def isHeader(data: ByteString) = DicomParsing.dicomInfo(data).isDefined
 
+        def readHeader(buffer: ByteString, assumeBigEndian: Boolean, explicitVR: Boolean): (Int, VR, Int, Int) = {
```
This method overlaps alot with `DicomPartFlow.readHeader`. Is there anything that perhaps can be generalized and moved to e.g. `DicomParsing`?

KarlSjostrand commented on src/main/scala/se/nimsa/dcm4che/streams/DicomFlows.scala:
```diff
@@ -62,34 +70,167 @@ object DicomFlows {
+            if (vr.headerLength == 8)
+              (tag, vr, 8, bytesToUShort(tagVr, 6, assumeBigEndian))
+            else {
+              (tag, vr, 12, bytesToInt(buffer, 8, assumeBigEndian))
```
In `DicomPartFlow.readHeader`, it is ensured that there are 12 bytes available before returning this. Should this be done here too?

KarlSjostrand commented on src/main/scala/se/nimsa/dcm4che/streams/DicomParsing.scala:
```diff
@@ -20,12 +20,31 @@ import akka.util.ByteString
   * Helper methods for parsing binary DICOM data.
   */
 trait DicomParsing {
+  val TRANSFER_SYNTAX_IMPLICIT_VR_LITTLE_ENDIAN = "1.2.840.10008.1.2"
```
All DICOM UIDs are defined in `org.dcm4che3.data.UID`. I would prefer using those.

KarlSjostrand commented on src/main/scala/se/nimsa/dcm4che/streams/DicomParsing.scala:
```diff
@@ -68,6 +115,10 @@ trait DicomParsing {
   }
 
   def isFileMetaInformation(tag: Int) = (tag & 0xFFFF0000) == 0x00020000
+  def isMediaStorageSOPClassUID(tag: Int) = tag == 0x00020002
```
Most tags are defined in `org.dcm4che3.data.Tag`. Replacing this method and the two below with code like `if (tag == Tag.MediaStorageSOPClassUID)` may be just as clear.

KarlSjostrand commented on src/test/scala/se/nimsa/dcm4che/streams/DicomFlowsTest.scala:
```diff
@@ -188,4 +189,272 @@ class DicomFlowsTest extends TestKit(ActorSystem("DicomAttributesSinkSpec")) wit
+
+  "The DICOM validation flow with contexts" should "buffer first 512 bytes" in {
+
+    val contexts = Seq(Context("1.2.840.10008.5.1.4.1.1.2", "1.2.840.10008.1.2.1"))
```
Use `org.dcm4che3.data.UID` constants.

kobmic commented:
Refactored PR:
- use constants from org.dcm4che  for Tags and UIDs
- moved validateFlow to own class, moved tests to own testclass
- refactored 'readHeader' method in validateFlow, split into 'readHeaderExplicit' and 'readHeaderImplicit' in DicomParsing

KarlSjostrand commented on src/main/scala/se/nimsa/dcm4che/streams/DicomParsing.scala:
```diff
@@ -19,13 +19,30 @@ package se.nimsa.dcm4che.streams
+      ImplicitVRLittleEndian
+    }
+  }
+  case class DataElement(tag: Int, vr: VR, length: Int, value: ByteString)
```
I thought a good terminology was that an `Attribute` consists of a `Header` and a `Value`. The dicom4che class `Attributes` is basically an aggregate of `Attribute`s, but I am in no way an expert on this. Is `DataElement` a better noun?

kobmic replied:
Changed from 'Attribute' to 'DataElement' after reading this:
ftp://dicom.nema.org/medical/DICOM/2013/output/chtml/part05/sect_3.10.html
Maybe 'Attribute' is better anyway?

KarlSjostrand replied:
Maybe it can be interpreted as that the `DataElement` is the actual data, whereas the `Attribute` is the interpreted information. The docs describe a data element as "An encoded Information Object Definition (IOD) Attribute".

KarlSjostrand commented on src/main/scala/se/nimsa/dcm4che/streams/DicomParsing.scala:
```diff
@@ -55,6 +72,82 @@ trait DicomParsing {
+  }
+
+
+  def valueWithoutPadding(value: ByteString) =
```
This method may only be used if you know the length of the input value, right? In other words, in cases where you know the value is padded? If not, you may be stripping a trailing zero that had some meaning. Just making sure I follow.

kobmic replied:
Hmm, only used in 'parseUIDDataElement', so we know the characters used in the value are "0"-"9", "." of Default Character Repertoire (according to the Standard). So if it's 00H it must be padding. Maybe should nest this method in  'parseUIDDataElement'?

KarlSjostrand commented on src/main/scala/se/nimsa/dcm4che/streams/DicomParsing.scala:
```diff
@@ -55,6 +72,82 @@ trait DicomParsing {
+    * @param assumeBigEndian true if big endian, false otherwise
+    * @return
+    */
+  def readHeaderExplicitVR(buffer: ByteString, assumeBigEndian: Boolean): Option[(Int, VR, Int, Int)] = {
```
Now that these methods have been refactored from DicomValidateFlow to DicomParsing, can they be used by DicomPartFlow as well to cut code?
